### ISSUE TITLE
Not all retained MQTT messages are received at NR full deploy

### ISSUE THREAD

**Issue by kbfifi** [node-red/node-red#2287]
- [ ] Node-RED version:
V0.17.5
- [ ] Node.js version:
v8.4.0
- [ ] npm version:
v6.9.0
- [ ] Platform/OS:
Linux DietPi 3.14.79+ #1 SMP PREEMPT Sun Jul 23 00:04:44 CEST 2017 aarch64 GNU/Linux
- [ ] Browser:
Firefox Quantum v69, Windows 10 Edu 1903

I use MQTT a lot in my home domotic projects. Amongst other Pi's I use the mentioned DietPi Odroid C2. On the board I run NR and I amongst things I store MQTT messages with settings at startup in global variables. 
Long story short: I miss retained messages at full deployment of the NR project. I tested with a debug node and I simply don't see all retained messages of a certain topic. If I check with a CLI client on another machine using the same broker, I see all of them. It used to work.
Any hints?

**Comment by knolleary** [comment]
Hi - thanks for raising an issue, we'd normally ask for this type of question to be raised on the forum where the wider community can help - as the issue template described.

When you say it 'used to work', what do you mean? Have you changed something?

It is worth pointing out that you are on an old version of Node-RED - 0.17.5 was released over 2 years ago.

**Comment by kbfifi** [comment]
Thanks for your quick response. As I tested the retained messages against the mosquitto client on the MQTT broker machine and the NR MQTT client did not give the same result I was thinking about a NR bug. But yes I should upgrade to a recent NR version to see if it's fixed meanwhile.
Also looked at: "Retained MQTT messages just partly received on restart/deploy #1950" That issue ended by being no longer reproducible. "seems to be dependent on the number of MQTT nodes using the same MQTT server" I have 2 NR Pi's and only 1 MQTT broker. Maybe it's related.
Now checking my issue again this evening shows: This issue not being reproducible... That makes it difficult to invetigate. Maybe the problem is with the broker but then again why would the broker client work just fine? I'll keep you informed if I find out more.

**Comment by kbfifi** [comment]
I finally got some time to fix my issue. I upgraded to NodeJS 13.5, Node-red 1.03
Problem remains!
But I think I found something now. If I subscribe on 'mytopic/#' and below this topic are subtopics say 'mytopic/subtopicA/valueB' they won't be received. However if I add another subscription 'mytopic/subtopicA/#' now valueB WILL be received.
For what it's worth: All topic are retained.
Hope it's of any value. Maybe it's a bug.

**Comment by dceejay** [comment]
A simple test seems to work, can you provide an example that doesn't ?

```
[{"id":"254ad429.6f77ec","type":"inject","z":"ea593186.4918d","name":"","topic":"mytopic/subtopicA/topicA","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":200,"y":900,"wires":[["9456484d.3608b8"]]},{"id":"c68cbe4d.43053","type":"inject","z":"ea593186.4918d","name":"","topic":"mytopic/subtopicB/topicA","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":200,"y":960,"wires":[["9456484d.3608b8"]]},{"id":"f484882e.90a2a8","type":"inject","z":"ea593186.4918d","name":"","topic":"mytopic/subtopicA/valueB","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":200,"y":1020,"wires":[["9456484d.3608b8"]]},{"id":"fa52b390.bae6e","type":"mqtt out","z":"ea593186.4918d","name":"","topic":"","qos":"","retain":"","broker":"4563ba42.326e34","x":610,"y":900,"wires":[]},{"id":"9c841f25.07c37","type":"mqtt in","z":"ea593186.4918d","name":"","topic":"mytopic/#","qos":"2","datatype":"auto","broker":"4563ba42.326e34","x":400,"y":1020,"wires":[["e5db5d06.35aa9"]]},{"id":"e5db5d06.35aa9","type":"debug","z":"ea593186.4918d","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","x":590,"y":1020,"wires":[]},{"id":"9456484d.3608b8","type":"change","z":"ea593186.4918d","name":"","rules":[{"t":"set","p":"retain","pt":"msg","to":"true","tot":"bool"}],"action":"","property":"","from":"","to":"","reg":false,"x":420,"y":900,"wires":[["fa52b390.bae6e"]]},{"id":"4563ba42.326e34","type":"mqtt-broker","z":"","name":"","broker":"test.mosquitto.org","port":"1883","clientid":"","usetls":false,"compatmode":false,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closeQos":"0","closePayload":"","willTopic":"","willQos":"0","willPayload":""}]
```

**Comment by kbfifi** [comment]
Hi dceejay, thanks for your reponse and checking upon it. I did some more investigation. I have have several devices with sub topics so I tried another one and that DID work. Question remains why this particular device didn't. So I just threw the MQTT node away and replaced it. Now it works! Exact same settings. One possible explanation is SD card data corruption on the Rasberry pi. I will never know because I cannot reproduce it anymore. 
Happy it works now. New board is on the way ;-)

**Comment by kbfifi** [comment]
Update:
- I disabled all flows
- created 1 flow with:
-  subscription timer/#
Shows same problem! Although I get 2 out of 20 timer/once/xx topics and the other 20 timer/daily/xx topics

When I only subscribe to timer/once/# I get all 20

**Comment by hardillb** [comment]
It is very unlikely that any fix would be provided on the 2.0.x branch, please try with 3.0.2

**Comment by kbfifi** [comment]
@hardillb Thanks for your comment. Good to upgrade anyway. 

I also found out that QOS of the client make a difference. I work with QOS 2 both for publishing and subscribing. Strange enough lowering to QOS to 0 in the mqtt subscribing node now shows more timer/once/x topics but also inconsitently. The short term workaround is to add subscribing node "timer/once/#" with QOS 0 addtionally to subscribing node "timer/#" with QOS.

After upgrade to latest 2.2.3 , in order to go to 3.0.2, I noticed every seems to work again! I see 2 explanations: 
1. Something went corrupt and is fixed by the update
2. There's a bug in node-red in some vague situation (impossible to reproduce of course)

Any other thoughts?

I guess reopen this issue in 3 years again ;-)

**Comment by kbfifi** [comment]
Oops cheered too early. I'm on 3.0.2 now. The issue is still there

**Comment by kbfifi** [comment]
Thanks for the suggestions. I checked with tcpdump and saw the correct subscription going to the broker but only timer/daily/xx coming back. So that proved that node-red works and likely the broker acts strange.
Then stopped the mosquitto service and I brutally removed the mosquitto.db file. Rebuild the retained messages and now everything works. Finally explained!
